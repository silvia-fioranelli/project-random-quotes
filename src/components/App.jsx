import React from 'react';
import { API_URL } from '../globals';
import QuoteBox from './QuoteBox';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quote: null,
      isError: false,
    };
    this.fetchData = this.fetchData.bind(this);
  }

  componentDidMount() {
    this.fetchData();
  }

  fetchData() {
    this.setState({
      quote: null,
      isError: false,
    });

    return fetch(API_URL)
      .then((res) => {
        if (!res.ok) {
          throw new Error('response not ok');
        }
        return res.json();
      })
      .then((data) => {
        this.setState({
          quote: data,
        });
      })
      .catch(() => {
        this.setState({
          isError: true,
        });
      });
  }

  render() {
    const { quote, isError } = this.state;

    return (
      <div
        className="d-flex align-items-center justify-content-center"
        style={{ height: '100vh' }}
      >
        <QuoteBox getQuote={this.fetchData} isError={isError} quote={quote} />
      </div>
    );
  }
}

export default App;
