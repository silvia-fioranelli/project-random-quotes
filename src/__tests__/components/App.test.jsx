import { shallow } from 'enzyme';
import App from '../../components/App';
import { API_URL } from '../../globals';

describe('App', () => {
  beforeEach(() => {
    fetch.mockImplementation(() => Promise.resolve({}));
  });

  describe('rendering', () => {
    test('component App should render', () => {
      const wrapper = shallow(<App />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    test('should call fetchData', () => {
      const wrapper = shallow(<App />);
      wrapper.instance().fetchData = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().fetchData).toBeCalledTimes(1);
    });
  });

  describe('fetchData', () => {
    test('should call API and set quotes', async () => {
      const data = {
        content: 'mock',
        author: 'mock',
      };
      const response = {
        ok: true,
        json: () => data,
      };
      const wrapper = shallow(<App />);
      fetch.mockClear();
      fetch.mockImplementationOnce(() => Promise.resolve(response));

      await wrapper.instance().fetchData();

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith(API_URL);

      expect(wrapper.instance().state.quote).toEqual(data);
    });

    test('should call API and set isError', async () => {
      const response = {
        ok: false,
      };
      const wrapper = shallow(<App />);
      fetch.mockClear();
      fetch.mockImplementationOnce(() => Promise.resolve(response));

      await wrapper.instance().fetchData();

      expect(fetch).toBeCalledTimes(1);
      expect(fetch).toBeCalledWith(API_URL);

      expect(wrapper.instance().state.isError).toEqual(true);
    });
  });
});
