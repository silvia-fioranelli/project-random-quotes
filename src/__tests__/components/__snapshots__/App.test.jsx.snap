// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering component App should render 1`] = `
<div
  className="d-flex align-items-center justify-content-center"
  style={
    Object {
      "height": "100vh",
    }
  }
>
  <QuoteBox
    getQuote={[Function]}
    isError={false}
    quote={null}
  />
</div>
`;
